master/buildbot/process/buildrequestdistributor.py

```diff
@@ -472,21 +473,28 @@ def _activityLoop(self):
 
         timer = metrics.Timer('BuildRequestDistributor._activityLoop()')
         timer.start()
-
+        pending_builders = []
         while True:
             yield self.activity_lock.acquire()
+            if not self.running:
+                self.activity_lock.release()
+                break
 
-            # lock pending_builders, pop an element from it, and release
-            yield self.pending_builders_lock.acquire()
+            if not pending_builders:
+                # lock pending_builders, pop an element from it, and release
+                yield self.pending_builders_lock.acquire()
 
-            # bail out if we shouldn't keep looping
-            if not self.running or not self._pending_builders:
+                # bail out if we shouldn't keep looping
+                if not self._pending_builders:
+                    self.pending_builders_lock.release()
+                    self.activity_lock.release()
+                    break
+                # take that builder list, and run it until the end
+                pending_builders = copy.copy(self._pending_builders)
```

Comment:
Why the copy?

Reply:
If you look at how locks are setup, it is not guaranteed that pending_builders will not be messed up while we are running the loop. 

Current implementation is not doing it, as it is playing with set and sorts. The locks looks useless as far as I can read. But as this stuff is pretty central, I prefer to be overcautious.